### PR TITLE
fix hack/local-up-cluster.sh apiserver not listening bug

### LINKED ISSUE #49728: local-up-cluster.sh run failed without apiserver listening
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
run hack/local-up-cluster.sh
apiserver start failed. with message no default route.

**What you expected to happen**:
start a local kubernetes cluster successfully.

**How to reproduce it (as minimally and precisely as possible)**:
run hack/local-up-cluster.sh or run hack/local-up-cluster.sh -O

**Anything else we need to know?**:
I suppose it is a bug.

**Environment**:
- Kubernetes version (use `kubectl version`):

Client Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.0-alpha.2.904+37215e67b8d490", GitCommit:"37215e67b8d4909a4f48faa19a3f2e97dc7565a9", GitTreeState:"clean", BuildDate:"2017-07-27T16:56:18Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.0-alpha.2.904+37215e67b8d490", GitCommit:"37215e67b8d4909a4f48faa19a3f2e97dc7565a9", GitTreeState:"clean", BuildDate:"2017-07-27T16:56:18Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration**:
no use

- OS (e.g. from /etc/os-release):
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

- Kernel (e.g. `uname -a`):
3.10.0-327.el7.x86_64 #1 SMP Thu Nov 19 22:10:57 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

- Install tools:
from source
- Others:
run local-up-cluster.sh in an intranet environment, with no internet connections.

### PR DESCRIPTION
**Pull request by LihuaWu**
**What this PR does / why we need it**:
This PR fix a potential bug int hack/local-up-cluster.sh and it helps developers to build local cluster properly

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fix #49728 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by LihuaWu** [comment]
the license has been signed

**Comment by sttts** [comment]
What exactly is the error you get without this PR (full log output please)?

**Comment by LihuaWu** [comment]
below is log from /tmp/kube-apiserver.log
2017-07-28 09:52:57.854224 I | proto: duplicate proto type registered: google.protobuf.Any
2017-07-28 09:52:57.854313 I | proto: duplicate proto type registered: google.protobuf.Duration
2017-07-28 09:52:57.854334 I | proto: duplicate proto type registered: google.protobuf.Timestamp
I0728 09:52:57.917582    8259 feature_gate.go:144] feature gates: map[PersistentLocalVolumes:false RotateKubeletClientCertificate:false Accelerators:false PodPriority:false APIResponseCompression:false AllAlpha:false TaintBasedEvictions:false LocalStorageCapacityIsolation:false ExperimentalCriticalPodAnnotation:false DynamicKubeletConfig:false RotateKubeletServerCertificate:false EnableEquivalenceClassCache:false DebugContainers:false DynamicVolumeProvisioning:false AdvancedAuditing:false]
I0728 09:52:57.917720    8259 server.go:114] Version: v1.8.0-alpha.2.904+37215e67b8d490
I0728 09:52:57.917750    8259 plugins.go:77] Registered admission plugin "AlwaysAdmit"
I0728 09:52:57.917757    8259 plugins.go:77] Registered admission plugin "AlwaysPullImages"
I0728 09:52:57.917761    8259 plugins.go:77] Registered admission plugin "LimitPodHardAntiAffinityTopology"
I0728 09:52:57.917766    8259 plugins.go:77] Registered admission plugin "DefaultTolerationSeconds"
I0728 09:52:57.917770    8259 plugins.go:77] Registered admission plugin "AlwaysDeny"
I0728 09:52:57.917775    8259 plugins.go:77] Registered admission plugin "DenyEscalatingExec"
I0728 09:52:57.917779    8259 plugins.go:77] Registered admission plugin "DenyExecOnPrivileged"
I0728 09:52:57.917783    8259 plugins.go:77] Registered admission plugin "OwnerReferencesPermissionEnforcement"
I0728 09:52:57.917789    8259 plugins.go:77] Registered admission plugin "ImagePolicyWebhook"
I0728 09:52:57.917794    8259 plugins.go:77] Registered admission plugin "Initializers"
I0728 09:52:57.917798    8259 plugins.go:77] Registered admission plugin "InitialResources"
I0728 09:52:57.917804    8259 plugins.go:77] Registered admission plugin "LimitRanger"
I0728 09:52:57.917809    8259 plugins.go:77] Registered admission plugin "NamespaceAutoProvision"
I0728 09:52:57.917815    8259 plugins.go:77] Registered admission plugin "NamespaceExists"
I0728 09:52:57.917821    8259 plugins.go:77] Registered admission plugin "NodeRestriction"
I0728 09:52:57.917828    8259 plugins.go:77] Registered admission plugin "PersistentVolumeLabel"
I0728 09:52:57.917838    8259 plugins.go:77] Registered admission plugin "PodNodeSelector"
I0728 09:52:57.917845    8259 plugins.go:77] Registered admission plugin "PodPreset"
I0728 09:52:57.917851    8259 plugins.go:77] Registered admission plugin "PodTolerationRestriction"
I0728 09:52:57.917856    8259 plugins.go:77] Registered admission plugin "ResourceQuota"
I0728 09:52:57.917861    8259 plugins.go:77] Registered admission plugin "PodSecurityPolicy"
I0728 09:52:57.917866    8259 plugins.go:77] Registered admission plugin "SecurityContextDeny"
I0728 09:52:57.917870    8259 plugins.go:77] Registered admission plugin "ServiceAccount"
I0728 09:52:57.917880    8259 plugins.go:77] Registered admission plugin "DefaultStorageClass"
I0728 09:52:57.917885    8259 plugins.go:77] Registered admission plugin "GenericAdmissionWebhook"
Error: Unable to find suitable network address.error='No default routes.'. Try to set the AdvertiseAddress directly or provide a valid BindAddress to fix this.

**Comment by sttts** [comment]
@LihuaWu can you please trace why `ChooseBindAddress` at `vendor/k8s.io/apimachinery/pkg/util/net/interface.go:383` does not work for your case? It should. I fear we are hiding another issue here. `adversize-address` should be automatic.

**Comment by LihuaWu** [comment]
@sttts sure. I will look at that and respond asap if i find something.

**Comment by LihuaWu** [comment]
@sttts I checked k8s.io/apimachinery/pkg/util/net/interface_test.go, and found my case.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/net/interface_test.go#L672

**Comment by sttts** [comment]
Any sensible way to solve this in your setup? Is it ipv6 only or something like that?

**Comment by LihuaWu** [comment]
No. Actually it has no internet connection since it is in an intranet enviroment. In my scenario, I access the internet through http_proxy. 

And, I don't think it is appropriate to force internet connection to run a local cluster.

**Comment by LihuaWu** [comment]
@sttts  I notice the comment said `Let the API server pick a default address when API_HOST_IP is set to 127.0.0.1`.
And May I know the reason why do we have to  pick a default address instead of just using the API_HOST_IP when it is set to 127.0.0.1? 
I am curious about it.

**Comment by sttts** [comment]
@LihuaWu compare the original PR https://github.com/kubernetes/kubernetes/issues/29374#issuecomment-234430562. `127.0.0.1` is even invalid as an endpoint address (b/c it's not routable in a docker container context). That was the reason that we have the if clause there. 

I still think that the apiserver heuristic for the external address is broken for your use-case. Maybe we even have to add a ADVERTISE_ADDRESS for your very use-case to local-up-cluster.sh. Feel free to do that if that fixes the script for your setup.

**Comment by LihuaWu** [comment]
@sttts I agree with you that the apiserver heuristic for the external address is broken for my use-case.  

And  the ADVERTISE_ADDRESS variable is useful to set an valid endpoint address. I  add  a new commit and it works in my case.  Below is details:

run with ADVERTISE_ADDRESS:  `ADVERTISE_ADDRESS="10.0.0.3" hack/local-up-cluster.sh  -O`

    cluster/kubectl.sh describe svc

    Name:            kubernetes
    Namespace:        default
    Labels:            component=apiserver
                provider=kubernetes
    Annotations:        <none>
    Selector:        <none>
    Type:            ClusterIP
    IP:            10.0.0.1
    Port:            https    443/TCP
    Endpoints:        10.0.0.3:6443
    Session Affinity:    ClientIP
    Events:            <none>

without ADVERTISE_ADDRESS:
It behaves as before,  the local-up-cluster.sh run failed.

**Comment by LihuaWu** [comment]
the fix should work when API_HOST_IP is set to 127.0.0.1 and the apiserver heuristic for the external address is broken, which is my case.

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -458,6 +459,10 @@ function start_apiserver {
     if [[ "${API_HOST_IP}" != "127.0.0.1" ]]; then
         advertise_address="--advertise_address=${API_HOST_IP}"
     fi
+    #Fix issue: https://github.com/kubernetes/kubernetes/issues/49728
```
Comment (sttts) [inline hack/local-up-cluster.sh]: Don't think we need the issue link here.

Comment (sttts) [inline hack/local-up-cluster.sh]: Put it into the commit message if anybody wants to find out.

**Comment by sttts** [comment]
@LihuaWu lgtm after removing the comment with the link. Just put it into the commit message.

**Comment by LihuaWu** [comment]
I push a new commit and thank you @sttts,  learnt a lot from you.

**Comment by sttts** [comment]
@LihuaWu congrats to your first Kubernetes contribution! 🚀

**Pull request merged**